src/index/CompressedBlockPrefiltering.h

```diff
@@ -0,0 +1,140 @@
+//  Copyright 2024, University of Freiburg,
+//                  Chair of Algorithms and Data Structures
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de>
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "global/Id.h"
+#include "global/ValueIdComparators.h"
+#include "index/CompressedRelation.h"
+
+// The compressed block metadata (see `CompressedRelation.h`) that we use to
+// filter out the non-relevant blocks by checking their content of
+// `firstTriple_` and `lastTriple_` (`PermutedTriple`)
+using BlockMetadata = CompressedBlockMetadata;
+
+namespace prefilterExpressions {
+//______________________________________________________________________________
+/*
+`PrefilterExpression` represents a base class for the following sub-classes that
+implement the block-filtering procedure for the specific relational + logical
+operations/expressions.
+
+Remark: We do not actually evaluate the respective SPARQL Expression. We only
+pre-filter w.r.t. blocks that contain relevant data for the actual evaluation of
+those expressions to make the evaluation procedure more efficient.
+
+The block-filtering is applied with the following operations:
+Relational Expressions - `<=`, `>=`, `<`, `>`, `==` and `!=`
+Logical Operations - `and` and `or`
+*/
+
+class PrefilterExpression {
+ public:
+  // The respective metadata to the blocks is expected to be provided in
+  // a sorted order (w.r.t. the relevant column).
+  virtual std::vector<BlockMetadata> evaluate(std::vector<BlockMetadata>& input,
+                                              size_t evaluationColumn) = 0;
```

Comment:
This function has some preconditions (especially: the input is sorted and unique, and all columns before the `evaluationColumn` are the same) as well as postconditions (the result is also sorted, a subset of the input, and unique).
You could make the current `evalute` function a private `evaluateImpl`, and then 
the public, non-virtual `evaluate` function checks the conditions before and after, and in the middle calls the virtual `evaluateImpl` for the actual computation.

Reply:
Thanks, good idea. I will implement it .

Reply:
The issue is then the implementation of `evaluateImpl`  for `LogicalExpressions` with `child_->evaluateImpl(input, column)` if `evaluateImpl` is private, or should i use `child_->evaluate(input, column)` where we check with each depth in the recursion?